File: sdks/typescript/src/filesystem.ts
export interface EntryInfo {
  name: string;
  isDir: boolean;
  path: string;
  size: number;
}

export class Filesystem {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private sandboxId: string,
    private token: string = "",
  ) {}

  private get headers(): Record<string, string> {
    if (this.token) return { "Authorization": `Bearer ${this.token}` };
    return this.apiKey ? { "X-API-Key": this.apiKey } : {};
  }

  async read(path: string): Promise<string> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      { headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to read ${path}: ${resp.status}`);
    return resp.text();
  }

  async readBytes(path: string): Promise<Uint8Array> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      { headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to read ${path}: ${resp.status}`);
    return new Uint8Array(await resp.arrayBuffer());
  }

  async write(path: string, content: string | Uint8Array): Promise<void> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      {
        method: "PUT",
        headers: this.headers,
        body: content as BodyInit,
      },
    );
    if (!resp.ok) throw new Error(`Failed to write ${path}: ${resp.status}`);
  }

  async list(path: string = "/"): Promise<EntryInfo[]> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files/list?path=${encodeURIComponent(path)}`,
      { headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to list ${path}: ${resp.status}`);
    const data = await resp.json();
    return data ?? [];
  }

  async makeDir(path: string): Promise<void> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files/mkdir?path=${encodeURIComponent(path)}`,
      { method: "POST", headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to mkdir ${path}: ${resp.status}`);
  }

  async remove(path: string): Promise<void> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      { method: "DELETE", headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to remove ${path}: ${resp.status}`);
  }

  /**
   * Returns the raw response body as a ReadableStream for large file downloads.
   * Avoids buffering the entire file in memory.
   */
  async readStream(path: string): Promise<ReadableStream<Uint8Array>> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      { headers: this.headers },
    );
    if (!resp.ok) throw new Error(`Failed to read ${path}: ${resp.status}`);
    if (!resp.body) throw new Error(`No response body for ${path}`);
    return resp.body;
  }

  /**
   * Streams a ReadableStream or Uint8Array as the request body for large file uploads.
   * Avoids buffering the entire file in memory when given a ReadableStream.
   */
  async writeStream(path: string, stream: ReadableStream<Uint8Array> | Uint8Array): Promise<void> {
    const opts: RequestInit & Record<string, unknown> = {
      method: "PUT",
      headers: this.headers,
      body: stream as BodyInit,
    };
    // duplex: "half" is required for ReadableStream bodies in Node.js fetch,
    // but must NOT be set for Uint8Array/Buffer bodies.
    if (stream instanceof ReadableStream) {
      opts.duplex = "half";
    }
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
      opts,
    );
    if (!resp.ok) throw new Error(`Failed to write ${path}: ${resp.status}`);
  }

  async exists(path: string): Promise<boolean> {
    try {
      const resp = await fetch(
        `${this.apiUrl}/sandboxes/${this.sandboxId}/files?path=${encodeURIComponent(path)}`,
        { headers: this.headers },
      );
      return resp.ok;
    } catch {
      return false;
    }
  }
}

File: sdks/rust/src/filesystem.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::error::{Error, Result};
use crate::sandbox::ClientCtx;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntryInfo {
    pub name: String,
    #[serde(rename = "isDir", default)]
    pub is_dir: bool,
    #[serde(default)]
    pub path: String,
    #[serde(default)]
    pub size: u64,
}

pub struct Filesystem {
    ctx: Arc<ClientCtx>,
    sandbox_id: String,
}

impl Filesystem {
    pub(crate) fn new(ctx: Arc<ClientCtx>, sandbox_id: String) -> Self {
        Self { ctx, sandbox_id }
    }

    fn url(&self) -> String {
        format!("{}/sandboxes/{}/files", self.ctx.api_url, self.sandbox_id)
    }

    pub async fn read(&self, path: &str) -> Result<String> {
        let resp = self
            .ctx
            .http
            .get(self.url())
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to read {}: {}", path, body),
            });
        }
        Ok(resp.text().await?)
    }

    pub async fn read_bytes(&self, path: &str) -> Result<Vec<u8>> {
        let resp = self
            .ctx
            .http
            .get(self.url())
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to read {}: {}", path, body),
            });
        }
        Ok(resp.bytes().await?.to_vec())
    }

    pub async fn write(&self, path: &str, content: impl Into<Vec<u8>>) -> Result<()> {
        let body: Vec<u8> = content.into();
        let resp = self
            .ctx
            .http
            .put(self.url())
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .body(body)
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to write {}: {}", path, body),
            });
        }
        Ok(())
    }

    pub async fn list(&self, path: &str) -> Result<Vec<EntryInfo>> {
        let resp = self
            .ctx
            .http
            .get(format!("{}/list", self.url()))
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to list {}: {}", path, body),
            });
        }
        let bytes = resp.bytes().await?;
        if bytes.is_empty() {
            return Ok(Vec::new());
        }
        // Server may return null when empty.
        let v: serde_json::Value = serde_json::from_slice(&bytes)?;
        if v.is_null() {
            return Ok(Vec::new());
        }
        Ok(serde_json::from_value(v)?)
    }

    pub async fn make_dir(&self, path: &str) -> Result<()> {
        let resp = self
            .ctx
            .http
            .post(format!("{}/mkdir", self.url()))
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to mkdir {}: {}", path, body),
            });
        }
        Ok(())
    }

    pub async fn remove(&self, path: &str) -> Result<()> {
        let resp = self
            .ctx
            .http
            .delete(self.url())
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Api {
                status: status.as_u16(),
                body: format!("failed to remove {}: {}", path, body),
            });
        }
        Ok(())
    }

    pub async fn exists(&self, path: &str) -> bool {
        match self
            .ctx
            .http
            .get(self.url())
            .headers(self.ctx.auth_only_headers())
            .query(&[("path", path)])
            .send()
            .await
        {
            Ok(resp) => resp.status().is_success(),
            Err(_) => false,
        }
    }
}
